accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+
+### Step 1: Python Interpreter
+
+Python interpreter in "vanilla version" should be packaged as usual in RPM, per OS. However, compiled statically for that specific OS version and architecture. This alone allows update Python interpreter independently from the currently running Salt Minion. The diffrence is that the Python interpreter is installed in an optional namespace, e.g. `/opt/python<VERSION>/bin/python<VERSION>` path. The very RPM package, however, is served by Uyuni Server repository, instead of default channels.
+
+This RPM package remains being vendor supported.
+
+### Step 2: Virtual Environment
+
+Once Python interpreter is delivered and installed on a managed system, Virtual Environment is created. It should contain Package Installer for Python (PIP) and Salt Minion required packages list.
+
+Salt Minion is no longer installed from the RPM package, but bootstrapped by PIP environment. However, the distributed package version that is on the PyPi server remains vendor supported.
+
+> Idea for audit: checksums of all the files in the Virtual Environment can be calculated and stored in the database that might run on PyPi container. This way it is possible to detect if the environment was tampered.
+
+### Step 3: Running Salt Minion
+
+Ready to go Salt Minion on the Virtual Environment is linked for the `systemd` as default and is started.
+
+### Next Steps: Updating Salt Minion
+
+In order to update Salt Minion to the latest Uyuni-supported version, Step 2 is repeated and new Virtual Environment is created nearby current one. Virtual Environment manager component should relink default environments shifting current to "previous" and "new" to "current", and then restart Salt Minion.
+
+### Restoring Failed Minion Update
+
+**Problem**
+New Salt Minion is completely broken and doesn't start anymore. This renders managed system no longer be able to be managed via Salt.
+
+**Rationale**
+Automatically restore and reconnect previous Salt Minion, once new Salt Minion upgrade went wrong and new Salt Minion is either fails to start at all or fails to communicate/respond or has any other major fatal issues.
+
+**Solution**
+One of the ideas is to switch Minions with a simple watchdog function, which has only one purpose: to watch status of some cookie file for a specific amount of time and restore previous minion, in case timeout is over. This function could be a part of the mechanism that takes care of managing Python Virtual Environments.
+
+E.g. if new Salt Minion restarted successfully and reported back, Salt Master can issue another command to "finalise upgrade" and so Salt Minion updates cookie file. In this case watchdog function will get a signal that the previous Virtual Environment restoration is no longer needed. Otherwise, previous Salt Minion is restored back.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
I still do not see the advantage of this idea.

* In an RPM / DEB based solution you need to create the packages, here too, and you need to convert them is something that can be installed via pip.

* In an RPM / DEB based solution, updating a single node is a matter of zypper up / apt-get update: this will take care of replacing the code, adjusting the migrations required for the new service, and restarting the service. In this idea you need to build a full infrastructure to create the venv, replace them, reconfiguring the services to point to the new venv, some watchdog to monitor the current status, etc, etc.

Willing to maintain a full Python for each arch / OS, I wonder how different is using a full RPM / DEB system where we provide our Python in /opt, and the minion and all the new libraries are living in /opt/pythonx.y/lib/site-packages/

Reply:
Advantage is not to create lots of dependencies on each arch, but all that delegate to PyPi. You not supposed to create lots of packageS, but one packagE: a completely static vanilla Python, which supposed to be enough to run venvs and pip, so the rest can come from PyPi local server. Unfortunately, in practice, however, it likely will need more than one static package for crypto, SSL... 😞

Reply:
I also thought it is quite obvious: if you want Python 3.8 on SLES11, running on S390, then probably it needs to be extra compiled there _anyway_, isn't it? 😉 It is straight-forward to [compile Python static](https://wiki.python.org/moin/BuildStatically) into one single RPM package. Basically the goal here is to have zero `Require`, something like `snap` image does.

Reply:
>  Unfortunately, in practice, however, it likely will need more than one static package for crypto, SSL... disappointed

Exactly.